### PR TITLE
Cassy's Media Ranker

### PR DESCRIPTION
**Pull request by cassyarchibald**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In my work model I have a method that can return the top ten works for a category. 
Describe how you approached testing that model method. What edge cases did you come up with? | was excited to try the things we learned this week and dove into coding rather than focusing on TDD for my models. I feel that my code suffered as a result and I realized that I had not written an outline of what to test for this method but had not actually written the tests. I'll try to do this by Monday night. Edge cases included what would happen if there were no works or if all works had no votes. 
What are `session` and `flash`? What is the difference between them? | Flash sends a one-time message from controllers to our views. It persists from one request to another. Once you set a flash notice, the user will see the notice the next time the page is rendered, then the browser forgets about it. Session keeps track of data throughout a users session (until they close the browser). It's similar to flash but the data is stored indefinitely.
Describe a controller filter you wrote. | I have a controller filter in the application controller to find the current user based on the session user id. 
What was one thing that you gained more clarity on through this assignment? | What types of methods are gained based on the relationship of your models. 
What is the Heroku URL of your deployed application? | https://cassys-media-ranker.herokuapp.com/welcome
Do you have any recommendations on how we could improve this project for the next cohort? | Perhaps spend a little more time on testing model methods. A take home worksheet would be great.

### CODE REVIEW

File: test/models/vote_test.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    @work = Work.first
+    @vote = Vote.new
+    @vote.user_id = @user.id
+    @vote.work_id = @work.id
+    @vote.save
+  end
+
+  describe 'validations' do
+    it 'can only have one vote per user' do
+      new_vote = Vote.new (
```
Comment (CheezItMan) [inline test/models/vote_test.rb]: One space between `Vote.new` and the `(`, is causing an error here.  Taking that out and all your tests pass, as written.  

Also you should have:  `expect(new_vote.valid?).must_equal false`

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  # Lets user login when they submit login form
+  post 'sessions/login', to: 'sessions#login', as: 'login'
+  # Brings up login form (get request)
+  get 'sessions/login', to: 'sessions#new', as: 'new_session'
+  # destroying the session
+  delete 'sessions/destroy', to: 'sessions#destroy', as: 'logout'
+
+  root 'welcome#index'
+
+  resources :welcome, only:[:index]
```
Comment (CheezItMan) [inline config/routes.rb:11]: Since you have the root path already routing to `Welcome#index` you don't need a route for `/welcome/`.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+  resources :works
+
+  put '/works/:id/upvote', to: 'works#upvote', as: 'upvote'
+
+# Only need user routes for new/show ?
+  resources :users, except:[:destroy, :create, :edit, :update]
```
Comment (CheezItMan) [inline config/routes.rb:18]: Do you use the `users/new` route?

This might be better as:

`  resources :users, only: [:index, :show]`

File: test/fixtures/votes.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+vote:
+  work_id: Work.first.id
```
Comment (CheezItMan) [inline test/fixtures/votes.yml:4]: You should instead use the nicknames for the other fixtures:

```yml
vote_one:
  work: rambo
  user: cassy
  date_created: 2018-10-08

vote_two:
...  
```

File: test/fixtures/users.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
```
Comment (CheezItMan) [inline test/fixtures/users.yml:2]: 👍

File: app/models/vote.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class Vote < ApplicationRecord
+  # A vote belongs to a single user, a single work
+  belongs_to :user
+  # Gives the option to do vote.user to get the user associated with the vote
+  belongs_to :work, counter_cache: true 
+
+  # I think this prevents voting more than once
+  # https://www.reddit.com/r/rails/comments/2su4ib/please_share_any_resources_related_to_modifying/
+  validates :user, uniqueness: { scope: :work, message: "You can only vote once for this work" }
+  validates :work, uniqueness: { scope: :user, message: "You can only vote once for this work" }
```
Comment (CheezItMan) [inline app/models/vote.rb:10]: The uniqueness validation is only needed once for the combination of two fields.

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+      expect(votes.length).must_be :>=, 1
+      votes.each do |vote|
+        expect(vote).must_be_instance_of Vote
+      end
+    end
+  end
+
+  describe 'custom methods' do
+    before do
```
Comment (CheezItMan) [inline test/models/work_test.rb:61]: I guess you ran outta time?

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<h2><%="User Summary: #{@user.name}" %></h2>
+<p>
+  <%= "Joined site: #{@user.date_joined}" %>
+</p>
+
+<!-- TODO Show votes  -->
+<section class="user-votes_container">
```
Comment (CheezItMan) [inline app/views/users/show.html.erb:7]: I see you ran outta time for this.  Just a note, this type of section where you show works would make an excellent use-case for a partial!

File: app/views/welcome/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+</h2>
+  <p><%= "#{@spotlight.votes.length} votes" %></p>
+</section>
+
+<div class="top_ten_container">
+
+  <section class="top_ten_list_container">
+    <h2>Top Movies</h2>
+    <ul class="list-group top_ten_list">
+      <% @movies.each do |movie| %>
```
Comment (CheezItMan) [inline app/views/welcome/index.html.erb:20]: Notice that you're listing these groups of works over and over again.  Maybe  a partial could DRY this up.

File: app/views/works/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,38 @@
+<h1> <%= action_name %> </h1>
+
+<%= form_with model: @work, class: "work-form"  do |f|%>
+  <p>Please provide the following information to save your media to our database:</p>
+  <div class="form-group">
+    <%= f.label :category %>
+    <select class="form_control">
+      <option>
+        <%= "book" %>
```
Comment (CheezItMan) [inline app/views/works/_form.html.erb:9]: Hardcoding the categories would work fine for now, but maybe you should have a constant or method you could use everywhere in your app to list the valid categories.  That might make an excellent class method for Work.

File: app/views/works/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,170 @@
+<div class="main_work_container">
+
+  <section class="media-table">
+    <h3><%="Movies"%></h3>
+    <table class="table">
+      <thead>
+        <tr>
+          <th>
+            <%= "Votes" %>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:9]: Notice again that this table is being repeated...

File: app/views/works/_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,38 @@
+<h1> <%= action_name %> </h1>
+
+<%= form_with model: @work, class: "work-form"  do |f|%>
+  <p>Please provide the following information to save your media to our database:</p>
+  <div class="form-group">
+    <%= f.label :category %>
+    <select class="form_control">
```
Comment (CheezItMan) [inline app/views/works/_form.html.erb:7]: You didn't put in `<%= f.select :category %>` here.  Did you not understand how to use the form generator for selects.

The documentation is [here](https://apidock.com/rails/ActionView/Helpers/FormBuilder/select).

File: app/models/work.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
+class Work < ApplicationRecord
+  # before_save :calculate_vote_total
+  # Many votes can be associated with a single work
+  has_many :votes
+  # has_many :users :through :votes
+
+  WORKS = ['movie', 'album', 'book']
```
Comment (CheezItMan) [inline app/models/work.rb:7]: This is a great constant, it could be wonderful when used in some of the views

File: app/models/work.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  validates :creator, presence: true
+  validates :publication_year, presence: true
+  validates :description, presence: true
+  validates :title, presence: true, uniqueness: true
+
+  def top_ten_list(category)
+    # Select all of category from work instances
+    work_by_category = Work.all.select { |work| work.category.downcase == "#{category}" }
+    # Return max by vote count for top 10
+    return work_by_category.max_by(10) { |work| work.votes.length }
```
Comment (CheezItMan) [inline app/models/work.rb:20]: I'm not sure that this also sorts the works, which would be useful

File: app/models/work.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  # def self.categories
```
Comment (CheezItMan) [inline app/models/work.rb:23]: This could have been a wonderful method...

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages
Comprehension questions	|  Check, I'll look into creating a short worksheet in the textbook curriculum.
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Pretty good routes, I left a comment or so.
Views are well-organized (DRY, use of semantic HTML, use of partials) |  Overall pretty good, you didn't finish them all, and you could have made better use of partials, it might have sped things up a bit.  
Errors are reported to the user |  Errors are reported, but the categories don't work!
Business logic lives in the models |  A little only the top-ten in a category.  You could have methods to return the lists sorted (entire list not just 10), and spotlight, and maybe upvote.  
Models are thoroughly tested, including relations, validations and any custom logic |  Looks like you didn't get finished with testing.  I did leave some comments in your test files and fixtures.  
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |  I can't add work due to a bug with your `select`.
**Wave 2 - Users and Votes** |
Users can log in and log out |  Check
The ID of the current user is stored in the session |  Check
A user cannot vote for the same media more than once | Check, but no flash notice is sent.
All media lists are ordered by vote count |  Check
Splash page contains a media spotlight |  Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users |  **Missing**
Individual user pages and the user list are present |  Check
**Optional - Styling** |
Bootstrap is used appropriately | For the most part yes, you could do with some improvements in the flash notice styling, but overall it's pretty good. 
Look and feel is similar to the original |  Relatively close given where you ran out of time.  
**Overall** |  Overall not bad, but it does look like you ran out of time and were unable to finish testing and some details in the views.  Take a look at my inline comments and let me know any questions you have.
